# IT2011 - Artificial Intelligence and Machine Learning

## Progress Review I
### Member 01 - Data Cleaning and Missing Data Handling

**Dataset:** Breast Cancer Wisconsin (Original)

**Main Responsibility:**
- Dataset investigation
- Missing data handling
- Duplicate investigation
- Data quality validation

In [2]:
%pip install pandas numpy matplotlib seaborn

   ---------------------------------------- 0.0/10.0 MB ? eta -:--:--
   - -------------------------------------- 0.3/10.0 MB ? eta -:--:--
   ---- ----------------------------------- 1.0/10.0 MB 4.2 MB/s eta 0:00:03
   -------- ------------------------------- 2.1/10.0 MB 4.2 MB/s eta 0:00:02
   ------------ --------------------------- 3.1/10.0 MB 4.3 MB/s eta 0:00:02
   --------------- ------------------------ 3.9/10.0 MB 4.5 MB/s eta 0:00:02
   ---------------- ----------------------- 4.2/10.0 MB 3.8 MB/s eta 0:00:02
   ----------------- ---------------------- 4.5/10.0 MB 3.4 MB/s eta 0:00:02
   ------------------ --------------------- 4.7/10.0 MB 3.1 MB/s eta 0:00:02
   -------------------- ------------------- 5.0/10.0 MB 2.8 MB/s eta 0:00:02
   ---------------------- ----------------- 5.5/10.0 MB 2.7 MB/s eta 0:00:02
   --------------------------- ------------ 6.8/10.0 MB 3.0 MB/s eta 0:00:02
   -------------------------------- ------- 8.1/10.0 MB 3.3 MB/s eta 0:00:01
   ----------

In [3]:
# ============================================
# STEP 1 - Import required libraries
# ============================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# ============================================
# STEP 2 - Load the original dataset
# ============================================

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.data"

df = pd.read_csv(url, header=None)

print("Dataset loaded successfully.")
print("Shape:", df.shape)

Dataset loaded successfully.
Shape: (699, 11)


In [5]:
# ============================================
# STEP 3 - Assign meaningful column names
# ============================================

columns = [
    "Sample_ID",
    "Clump_Thickness",
    "Uniformity_Cell_Size",
    "Uniformity_Cell_Shape",
    "Marginal_Adhesion",
    "Single_Epithelial_Cell_Size",
    "Bare_Nuclei",
    "Bland_Chromatin",
    "Normal_Nucleoli",
    "Mitoses",
    "Class"
]

df.columns = columns

print("Column names assigned successfully.")
print(df.columns.tolist())

Column names assigned successfully.
['Sample_ID', 'Clump_Thickness', 'Uniformity_Cell_Size', 'Uniformity_Cell_Shape', 'Marginal_Adhesion', 'Single_Epithelial_Cell_Size', 'Bare_Nuclei', 'Bland_Chromatin', 'Normal_Nucleoli', 'Mitoses', 'Class']


In [6]:
# ============================================
# STEP 4 - Inspect the dataset
# ============================================

print("First 5 rows:")
display(df.head())

print("\nDataset Shape:")
print(df.shape)

print("\nData Types:")
print(df.dtypes)

First 5 rows:


,Sample_ID,Clump_Thickness,Uniformity_Cell_Size,Uniformity_Cell_Shape,Marginal_Adhesion,Single_Epithelial_Cell_Size,Bare_Nuclei,Bland_Chromatin,Normal_Nucleoli,Mitoses,Class
0,1000025,5,1,1,1,2,1,3,1,1,2
1,1002945,5,4,4,5,7,10,3,2,1,2
2,1015425,3,1,1,1,2,2,3,1,1,2
3,1016277,6,8,8,1,3,4,3,7,1,2
4,1017023,4,1,1,3,2,1,3,1,1,2



Dataset Shape:
(699, 11)

Data Types:
Sample_ID                      int64
Clump_Thickness                int64
Uniformity_Cell_Size           int64
Uniformity_Cell_Shape          int64
Marginal_Adhesion              int64
Single_Epithelial_Cell_Size    int64
Bare_Nuclei                      str
Bland_Chromatin                int64
Normal_Nucleoli                int64
Mitoses                        int64
Class                          int64
dtype: object


In [7]:
# ============================================
# STEP 5 - Investigate missing values
# ============================================

print("Number of '?' values in each column:")
print((df == "?").sum())

Number of '?' values in each column:
Sample_ID                       0
Clump_Thickness                 0
Uniformity_Cell_Size            0
Uniformity_Cell_Shape           0
Marginal_Adhesion               0
Single_Epithelial_Cell_Size     0
Bare_Nuclei                    16
Bland_Chromatin                 0
Normal_Nucleoli                 0
Mitoses                         0
Class                           0
dtype: int64


In [8]:
# ============================================
# STEP 6 - Convert '?' values to NaN
# ============================================

df["Bare_Nuclei"] = df["Bare_Nuclei"].replace("?", np.nan)

df["Bare_Nuclei"] = pd.to_numeric(df["Bare_Nuclei"])

print("Data type of Bare_Nuclei:")
print(df["Bare_Nuclei"].dtype)

print("\nMissing values after conversion:")
print(df.isnull().sum())

Data type of Bare_Nuclei:
float64

Missing values after conversion:
Sample_ID                       0
Clump_Thickness                 0
Uniformity_Cell_Size            0
Uniformity_Cell_Shape           0
Marginal_Adhesion               0
Single_Epithelial_Cell_Size     0
Bare_Nuclei                    16
Bland_Chromatin                 0
Normal_Nucleoli                 0
Mitoses                         0
Class                           0
dtype: int64


In [9]:
# ============================================
# STEP 7 - Analyze rows with missing Bare_Nuclei
# ============================================

missing_rows = df[df["Bare_Nuclei"].isnull()]

print("Number of rows with missing Bare_Nuclei:")
print(len(missing_rows))

print("\nClass distribution among missing rows:")
print(missing_rows["Class"].value_counts())

missing_percentage = (
    df["Bare_Nuclei"].isnull().sum() / len(df)
) * 100

print("\nPercentage of missing Bare_Nuclei values:")
print(f"{missing_percentage:.2f}%")

Number of rows with missing Bare_Nuclei:
16

Class distribution among missing rows:
Class
2    14
4     2
Name: count, dtype: int64

Percentage of missing Bare_Nuclei values:
2.29%


In [10]:
# ============================================
# STEP 8 - Analyze Bare_Nuclei distribution
# ============================================

print("Descriptive statistics for Bare_Nuclei:")
print(df["Bare_Nuclei"].describe())

print("\nFrequency of Bare_Nuclei values:")
print(df["Bare_Nuclei"].value_counts().sort_index())

Descriptive statistics for Bare_Nuclei:
count    683.000000
mean       3.544656
std        3.643857
min        1.000000
25%        1.000000
50%        1.000000
75%        6.000000
max       10.000000
Name: Bare_Nuclei, dtype: float64

Frequency of Bare_Nuclei values:
Bare_Nuclei
1.0     402
2.0      30
3.0      28
4.0      19
5.0      30
6.0       4
7.0       8
8.0      21
9.0       9
10.0    132
Name: count, dtype: int64


In [11]:
# ============================================
# STEP 9 - Handle missing values using median
# ============================================

bare_nuclei_median = df["Bare_Nuclei"].median()

print("Median of Bare_Nuclei:", bare_nuclei_median)

df["Bare_Nuclei"] = df["Bare_Nuclei"].fillna(
    bare_nuclei_median
)

print("\nMissing values after median imputation:")
print(df.isnull().sum())

Median of Bare_Nuclei: 1.0

Missing values after median imputation:
Sample_ID                      0
Clump_Thickness                0
Uniformity_Cell_Size           0
Uniformity_Cell_Shape          0
Marginal_Adhesion              0
Single_Epithelial_Cell_Size    0
Bare_Nuclei                    0
Bland_Chromatin                0
Normal_Nucleoli                0
Mitoses                        0
Class                          0
dtype: int64


In [12]:
# ============================================
# STEP 10 - Investigate exact duplicate rows
# ============================================

# Create a temporary copy for duplicate checking
df_duplicate_check = df.copy()

# Check completely identical rows
exact_duplicates = df_duplicate_check.duplicated().sum()

print("Number of exact duplicate rows:")
print(exact_duplicates)

Number of exact duplicate rows:
9


In [13]:
# ============================================
# STEP 11 - Remove exact duplicate observations
# ============================================

before_duplicates = len(df)

df = df.drop_duplicates().reset_index(drop=True)

after_duplicates = len(df)

print("Rows before removing duplicates:", before_duplicates)
print("Rows after removing duplicates:", after_duplicates)
print("Exact duplicate rows remaining:", df.duplicated().sum())

Rows before removing duplicates: 699
Rows after removing duplicates: 690
Exact duplicate rows remaining: 0


In [14]:
# ============================================
# STEP 12 - Analyze Bare_Nuclei distribution
# ============================================

print("Descriptive statistics for Bare_Nuclei:")
print(df["Bare_Nuclei"].describe())

print("\nMedian of Bare_Nuclei:")
print(df["Bare_Nuclei"].median())

Descriptive statistics for Bare_Nuclei:
count    690.000000
mean       3.482609
std        3.617064
min        1.000000
25%        1.000000
50%        1.000000
75%        5.000000
max       10.000000
Name: Bare_Nuclei, dtype: float64

Median of Bare_Nuclei:
1.0


In [15]:
# ============================================
# STEP 13 - Final data quality check
# ============================================

print("Final dataset shape:")
print(df.shape)

print("\nMissing values:")
print(df.isnull().sum())

print("\nExact duplicate rows:")
print(df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

Final dataset shape:
(690, 11)

Missing values:
Sample_ID                      0
Clump_Thickness                0
Uniformity_Cell_Size           0
Uniformity_Cell_Shape          0
Marginal_Adhesion              0
Single_Epithelial_Cell_Size    0
Bare_Nuclei                    0
Bland_Chromatin                0
Normal_Nucleoli                0
Mitoses                        0
Class                          0
dtype: int64

Exact duplicate rows:
0

Data types:
Sample_ID                        int64
Clump_Thickness                  int64
Uniformity_Cell_Size             int64
Uniformity_Cell_Shape            int64
Marginal_Adhesion                int64
Single_Epithelial_Cell_Size      int64
Bare_Nuclei                    float64
Bland_Chromatin                  int64
Normal_Nucleoli                  int64
Mitoses                          int64
Class                            int64
dtype: object


## Data Cleaning Summary

- The original dataset contained 699 observations and 11 columns.
- 16 missing values were identified in the Bare_Nuclei feature.
- The missing values were represented by `?`.
- `?` values were converted to NaN and Bare_Nuclei was converted to numeric.
- Exact duplicate observations were investigated and removed.
- 8 exact duplicate observations were removed.
- The dataset was reduced from 699 to 691 observations.
- Bare_Nuclei missing values were handled using median imputation.
- The final cleaned dataset contained no missing values and no exact duplicate rows.

In [16]:
# ============================================
# STEP 14 - Save cleaned outputs
# ============================================

df.to_csv(
    "results/outputs/cleaned_dataset.csv",
    index=False
)

missing_summary = df.isnull().sum().reset_index()
missing_summary.columns = ["Column", "Missing_Values"]

missing_summary.to_csv(
    "results/outputs/missing_value_summary.csv",
    index=False
)

print("Cleaned dataset and missing-value summary saved successfully.")

Cleaned dataset and missing-value summary saved successfully.
